File: dotnet/ThirtyFour.UI/Window/UrlMatcher.cs
using OpenQA.Selenium;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ThirtyFour.UI.Window
{
    /// <summary>
    /// The UrlMatcher class matches and switches to the first popup window it finds which matches the provided url
    /// </summary>
    public class UrlMatcher : IMatcher
    {
        private string url;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="url">The url of the window to match</param>
        public UrlMatcher(string url)
        {
            this.url = url;
        }

        /// <summary>
        /// Switches to the first popup window found which matches the provided url
        /// </summary>
        /// <param name="driver">The IWebDriver instance</param>
        /// <param name="timeoutInSeconds">The number of seconds to keep trying to find the matching popup window</param>
        public void MatchWindow(IWebDriver driver, double timeoutInSeconds)
        {
            bool found = Utils.RetryUntilTimeout(TimeSpan.FromSeconds(timeoutInSeconds), TimeSpan.FromMilliseconds(100), () => {
                IReadOnlyCollection<String> windows = driver.WindowHandles;

                foreach (String window in windows)
                {
                    driver.SwitchTo().Window(window);
                    if (driver.Url.Equals(this.url))
                    {
                        return true;
                    }
                }
                return false;
            });

            if (!found)
            {
                throw new Exception("Unable to find Window");
            }
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/window/UrlMatcher.java
package com.github.danhumphrey.thirtyfour.ui.window;

import java.util.Set;
import java.util.concurrent.Callable;

import org.openqa.selenium.WebDriver;

import com.github.danhumphrey.thirtyfour.ui.util.Utils;

/**
 * The UrlMatcher matches the first browser window it finds with a matching title
 *
 */
public class UrlMatcher implements Matcher {
	
	private String url;
	
	/**
	 * Constructor
	 * @param title the title of the window you want to match
	 */
	public UrlMatcher(String url) {
		this.url = url;
	}
	
	public class MatcherTask implements Callable<Boolean> {
	    
		private WebDriver driver;
		
		public MatcherTask(WebDriver driver) {
			this.driver = driver;
		}
		
		public Boolean call() throws InterruptedException {
			
			String currentHandle = driver.getWindowHandle();
			
	    	Set<String> handles = driver.getWindowHandles();
	    	for(String handle : handles) {
	    		
	    		driver.switchTo().window(handle);
	    		if(driver.getCurrentUrl().equals(url)) {
	    			return true;
	    		}
	    		driver.switchTo().window(currentHandle);
	    	}
	    	return false;
	    }
	}
	
	@Override
	public void matchWindow(WebDriver driver, long timeoutInMilliseconds) throws Exception {
		
		boolean found = Utils.retryUntilTimeout(timeoutInMilliseconds, 100, new MatcherTask(driver));
		
		if (!found) {
            throw new Exception("Unable to find a matching window");
        }
			
	}

}